util/build_configs/cray-internal/generate-modulefile.bash

```diff
@@ -76,7 +76,35 @@ if { [string match aarch64 $CHPL_HOST_ARCH] } {
     # ARM-based CPU, 2018-06-08
 } elseif { [string match x86_64 $CHPL_HOST_ARCH] } {
     # Legacy Cray-XC/XE
-    module load cray-mpich
+
+    # Load/unload cray-mpich if not previously loaded
+
+    set mpichLoaded [string match "*PE_MPICH*" $env(PE_PKGCONFIG_PRODUCTS)]
+    # Logic for mpich is split into loading and unloading phases
+    if { !([is-loaded chapel] == 1) }  {
+        # Loading chapel
+
+        if {$mpichLoaded} {
+            setenv CHPL_MODULE_KEEP_MPICH 1
+        }
+
+        if {! $mpichLoaded} {
+            module load cray-mpich
+        }
+    } else {
+        # Unloading or reloading chapel
+
+        if {$mpichLoaded} {
+            # Was hugepages already loaded before we loaded chapel?
```

Comment:
This and the next two comments should be rephrased to talk about cray-mpich instead of hugepages.  Actually, the last one can probably be removed altogether, since it seems just to be clarifying which hugepages module we'll unload.